File: src/pved_ptau217/validation.py
"""Input validation that fails early on ambiguous analysis data."""

from __future__ import annotations

from collections.abc import Iterable

import numpy as np
import pandas as pd


def require_columns(frame: pd.DataFrame, columns: Iterable[str], table_name: str) -> None:
    required = list(dict.fromkeys(columns))
    missing = [column for column in required if column not in frame.columns]
    if missing:
        raise ValueError(f"{table_name} is missing required columns: {missing}")


def require_unique(frame: pd.DataFrame, column: str, table_name: str) -> None:
    if frame[column].isna().any():
        raise ValueError(f"{table_name}.{column} contains missing values.")
    duplicated = frame[column].duplicated(keep=False)
    if duplicated.any():
        raise ValueError(
            f"{table_name}.{column} must be unique; "
            f"{int(duplicated.sum())} rows are duplicated."
        )


def require_valid_dates(
    frame: pd.DataFrame,
    columns: Iterable[str],
    table_name: str,
    *,
    allow_missing: bool,
) -> pd.DataFrame:
    checked = frame.copy()
    for column in columns:
        parsed = pd.to_datetime(checked[column], errors="coerce")
        invalid = checked[column].notna() & parsed.isna()
        if invalid.any():
            raise ValueError(
                f"{table_name}.{column} contains {int(invalid.sum())} invalid dates."
            )
        if not allow_missing and parsed.isna().any():
            raise ValueError(f"{table_name}.{column} contains missing dates.")
        checked[column] = parsed
    return checked


def require_positive(
    frame: pd.DataFrame,
    column: str,
    table_name: str,
    *,
    allow_missing: bool = True,
) -> None:
    values = pd.to_numeric(frame[column], errors="coerce")
    invalid_text = frame[column].notna() & values.isna()
    if invalid_text.any():
        raise ValueError(f"{table_name}.{column} contains nonnumeric values.")
    if not allow_missing and values.isna().any():
        raise ValueError(f"{table_name}.{column} contains missing values.")
    nonpositive = values.notna() & values.le(0)
    if nonpositive.any():
        raise ValueError(
            f"{table_name}.{column} contains {int(nonpositive.sum())} "
            "non-positive values; log10 transformation is undefined."
        )


def require_finite_variation(frame: pd.DataFrame, columns: Iterable[str]) -> None:
    for column in columns:
        values = pd.to_numeric(frame[column], errors="coerce")
        finite = values[np.isfinite(values)]
        if finite.nunique() < 2:
            raise ValueError(f"{column} has fewer than two finite values.")

File: R/validation.R
require_columns <- function(data, columns, table_name) {
  missing <- setdiff(columns, names(data))
  if (length(missing) > 0) {
    stop(
      sprintf("%s is missing required columns: %s", table_name, paste(missing, collapse = ", ")),
      call. = FALSE
    )
  }
  invisible(TRUE)
}

require_unique <- function(data, column, table_name) {
  duplicates <- data[[column]][duplicated(data[[column]])]
  if (length(duplicates) > 0) {
    stop(sprintf("%s has duplicate participant rows in '%s'.", table_name, column), call. = FALSE)
  }
  invisible(TRUE)
}

parse_required_dates <- function(data, columns, table_name, allow_missing = FALSE) {
  out <- data
  for (column in columns) {
    parsed <- as.Date(out[[column]])
    if (!allow_missing && any(is.na(parsed))) {
      stop(sprintf("%s has missing or invalid dates in '%s'.", table_name, column), call. = FALSE)
    }
    bad <- !is.na(out[[column]]) & is.na(parsed)
    if (any(bad)) {
      stop(sprintf("%s has invalid dates in '%s'.", table_name, column), call. = FALSE)
    }
    out[[column]] <- parsed
  }
  out
}

require_positive <- function(data, column, table_name, allow_missing = TRUE) {
  values <- suppressWarnings(as.numeric(data[[column]]))
  checked <- if (allow_missing) values[!is.na(values)] else values
  if (any(is.na(checked)) || any(checked <= 0)) {
    stop(sprintf("%s requires positive finite values in '%s'.", table_name, column), call. = FALSE)
  }
  invisible(TRUE)
}

require_finite_variation <- function(data, columns, table_name = "model table") {
  for (column in columns) {
    values <- suppressWarnings(as.numeric(data[[column]]))
    if (any(!is.finite(values))) {
      stop(sprintf("%s has non-finite values in '%s'.", table_name, column), call. = FALSE)
    }
    if (stats::sd(values) == 0 || is.na(stats::sd(values))) {
      stop(sprintf("%s has no variation in '%s'.", table_name, column), call. = FALSE)
    }
  }
  invisible(TRUE)
}

write_csv_public <- function(data, path) {
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  utils::write.csv(data, path, row.names = FALSE, na = "")
}
